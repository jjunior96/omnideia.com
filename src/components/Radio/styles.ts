import { lighten } from 'polished';
import styled, { css } from 'styled-components';

export const Container = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;

    & + div {
      margin-top: ${theme.spacings.xxxsmall};
    }
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xsmall};
    color: ${theme.colors.gray_300};
    margin: 0;
    padding: 0;
    line-height: 1;
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: center;
    appearance: none;

    border: 0.2rem solid ${theme.colors.purple_300};

    width: 1.8rem;
    height: 1.8rem;

    cursor: pointer;

    background: transparent;

    border-radius: 50%;

    color: ${theme.colors.black};

    margin-right: ${theme.spacings.xxxsmall};
    outline: none;

    transition: background ${theme.transition.fast};

    &:focus {
      outline: 3px solid ${lighten(0.2, theme.colors.purple_300)};
    }

    &:before {
      content: '';
      width: 0.8rem;
      height: 0.8rem;

      border-radius: 50%;

      background: ${theme.colors.purple_300};
      transition: opacity ${theme.transition.fast};

      opacity: 0;

      position: absolute;
    }

    &:checked {
      &:before {
        opacity: 1;
      }
    }
  `}
`;
