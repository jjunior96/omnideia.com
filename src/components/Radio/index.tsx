import { InputHTMLAttributes } from 'react';

import * as S from './styles';

type RadioValue = string | number | ReadonlyArray<string>;

export type RadioProps = {
  label?: string;
  onCheck?: (value: RadioValue) => void;
  value?: RadioValue;
} & InputHTMLAttributes<HTMLInputElement>;

const Radio = ({ label, value, onCheck, ...rest }: RadioProps) => {
  const onChange = () => {
    !!onCheck && onCheck(value!);
  };

  return (
    <S.Container>
      <S.Input
        id={label}
        type="radio"
        value={value}
        onChange={onChange}
        {...rest}
      />
      {!!label && <S.Label htmlFor={label}>{label}</S.Label>}
    </S.Container>
  );
};

export default Radio;
