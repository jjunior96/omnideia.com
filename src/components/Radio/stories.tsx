import { Story, Meta } from '@storybook/react/types-6-0';

import Radio, { RadioProps } from '.';

export default {
  title: 'Form/Radio',
  component: Radio,
  decorators: [
    (Story) => (
      <div style={{ width: '40rem', margin: '0 auto', padding: '1rem 0 1rem' }}>
        <Story />
      </div>
    )
  ],
  argTypes: {
    onCheck: {
      action: 'checked'
    },
    value: {
      type: ''
    }
  }
} as Meta;

export const Default: Story<RadioProps> = (args) => (
  <>
    <Radio
      label="Primeiro"
      name="nome"
      value="primeiro"
      defaultChecked
      {...args}
    />

    <Radio label="Segundo" name="nome" value="segundo" {...args} />

    <Radio label="Terceiro" name="nome" value="terceiro" {...args} />
  </>
);
