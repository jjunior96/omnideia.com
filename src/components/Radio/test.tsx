import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/helpers';

import Radio from '.';

describe('<Radio />', () => {
  it('should render the Radio component', () => {
    const { container } = renderWithTheme(<Radio label="Radio" value="any" />);

    expect(screen.getByRole('radio', { name: /Radio/i })).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the Radio without label', () => {
    renderWithTheme(<Radio value="any" />);

    expect(screen.queryByLabelText(/Radio/i)).not.toBeInTheDocument();
  });

  it('should dispatch onCheck when user click on element', async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Radio name="checkbox" onCheck={onCheck} />);

    // NÃ£o deve chamar a funcao onCheck
    expect(onCheck).not.toHaveBeenCalled();

    // Ao clicar no elemento, deve chamar a funcao onCheck
    userEvent.click(screen.getByRole('radio'));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });
  });

  it('should be accessible with tab', () => {
    renderWithTheme(<Radio name="Radio" label="Check" />);

    // No inicio, o focus deve estar no body da pagina
    expect(document.body).toHaveFocus();

    // Ao apertar em tab, o focus deve estar no checkbox
    userEvent.tab();

    expect(screen.getByRole('radio')).toHaveFocus();
  });
});
