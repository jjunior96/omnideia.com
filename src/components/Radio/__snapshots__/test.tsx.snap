// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render the Radio component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 + div {
  margin-top: 0.8rem;
}

.c2 {
  font-size: 1.6rem;
  color: #3E606F;
  margin: 0;
  padding: 0;
  line-height: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0.2rem solid #5855E9;
  width: 1.8rem;
  height: 1.8rem;
  cursor: pointer;
  background: transparent;
  border-radius: 50%;
  color: #0D0D0D;
  margin-right: 0.8rem;
  outline: none;
  -webkit-transition: background 0.1s ease-in-out;
  transition: background 0.1s ease-in-out;
}

.c1:focus {
  outline: 3px solid #b1aff5;
}

.c1:before {
  content: '';
  width: 0.8rem;
  height: 0.8rem;
  border-radius: 50%;
  background: #5855E9;
  -webkit-transition: opacity 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out;
  opacity: 0;
  position: absolute;
}

.c1:checked:before {
  opacity: 1;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="Radio"
    type="radio"
    value="any"
  />
  <label
    class="c2"
    for="Radio"
  >
    Radio
  </label>
</div>
`;
